### ISSUE TITLE
Make Fleks multi platform

### ISSUE THREAD

**Issue by Quillraven** [Quillraven/Fleks#2]
I am not an expert when it comes to Kotlin MPP but I guess it should not be super difficult to convert the current code to MPP?
Would be great if we can achieve this to support any backend.

Please share good articles/tutorials about MPP or create a PR yourself! Would be much appreciated.

**Comment by jobe-m** [comment]
Just wanted you to know that I started to work on porting Fleks to Korge :)
I stumbed over your project while reading in the Korge Discord and got pushed by v79.
Since I am currently interested in ECS a well so I desided to abandon my own static implementation and try to get Fleks working instead.
I read forward and backwards though the code already a like it very much. Especially the nice and elegant API. So I hope that "removing" all the advanced Java extentions (especially Annotations) I do not ruin the nice API :)
Here is the progress so far: https://github.com/korlibs/korge-next/issues/433

Thanks for doing Fleks!

**Comment by Quillraven** [comment]
@jobe-m: wow, so cool! thank you for your help.

Unfortunately, I have no experience still with KMM. Someone mentioned in some discussions that I should use annotation processing instead of reflection stuff. I am not sure if the reflection I am doing right now is causing problems but if it does then maybe annotation processing is the solution for it? 

Regarding annotations itself: It should work for JVM/Kotlin Native/Android backends, right? For JS I guess there should also be a solution because e.g. Angular uses annotations from what I know. The last backend is ios where I am not sure.

Anyway, if it is a show stopper for KMM then I am up for a discussion of course! But that for sure means we need to change the way how families are defined for IteratingSystems and how to resolve multiple dependencies of the same type. That's the only two locations where Fleks uses annotations at the moment.

**Comment by jobe-m** [comment]
I guess that I found already a good replacement for the java reflection code in Fleks. It was used to create instances from class constructors. But this feature is not available in Kotlin JS and that backend is frequently used when writing Korge Apps. So passing constructor functions as factory method does the trick to replace resp. reduce the reflection code to that what JS supports.

Annotations are another thing. While in Kotlin they are not used widely. It looks like they are quite common in Java code. (I am not a professional java dev...) So, here I need to read a bit about how annotations are working and then check which annotation features are actually supported in Kotlin JS and Kotlin Native.

I have today one day off. Just sitting beside my kid. Trying to persuade him to do his homework and reading my Kotlin Kompendium side by side ;)

**Comment by jobe-m** [comment]
Update - What I have now working:

- creation of systems
  -> with passing constructor method instead of just specifying the type
- inject external dependencies
  -> Since I cannot use much of reflection code with Kotlin JS/native I used here a singleton "Inject" object. It was the easiest and cleanes way of doing that. ;)
- inject componentMapper
  -> Here I also used the singleton Inject object.
- registering (and creating) components
  -> Here I needed to add a function "component(...)" to the WorldConfiguration which adds / registers a component to the world. This was needed again for passing the constructor method. I did not figured it out how I could create component objects when initializing the Family objects.
- Family
  -> Here comes the biggest pitfall. Unfortunately I needed to drop Annotations at all for specifying families of components. I just didn't get it working to access the (class) annotations when setting up the system objects. It looks like because Kotlin native (and JS) is so limited regarding to Annotations. I also did not find any solution or suggestions on Stackoverflow, etc. I added arrays of components for allOf, noneOf and anyOf as constructor properties instead to get it somehow working. (and not changing too much how it worked before...)
- adding entity
  -> this worked mostly out of the box

Currently I am working on adding ComponentListeners. It looks like it will be similar to adding Systems.
When I finished that I will update the main readme code snippets to show the main differences in the API.

**Comment by Quillraven** [comment]
@jobe-m: sounds great! Really appreciate your help 👍 

I have one question though: is the main problem with annotations the Java reflection part? Because by accident I saw on the weekend that `kotlinx-serialization` is also using annotations and that one should be multi platform compatible from what I read.

They have e.g. a `@Serializable` annotation and from what I read it uses a plugin to automatically create a `serializer` function for the companion object of the class who gets annotated with that.

I am not an expert either when it comes to annotation processors but I guess that's the reason why you have to add a gradle plugin as well when using kotlinx serialization so that plugin can create the code magically in the background for the users.

I know that @czybzy always mentioned annotation processing as a solution for some problems that I faced during the development of Fleks. Maybe it is time to investigate this now in more detail when it comes to KMM support? What do you think?

I anyway would plan that as a `2.0` version as it is a major change and big milestone for the project. I am not sure if this is possible but what we could do is:
- Create a plugin for the family annotations. It could maybe create like a `family` field for a system class that returns the configured family by the user. Maybe the plugin can simply create such an instance somehow.
- Introduce a new `@Component` annotation that automatically adds a companion object field of type Int that is unique for each component. It could also add a `newInstance` companion object function that takes no argument and returns a new instance of a component respecting the default value assignments. That way we could maybe even remove most of the ComponentMapper logic.
- Maybe a `@System` annotation could then also get rid off the system dependency injection via reflection.

----

I think for now, please continue with what you have so we have a working solution for KMM available, which will be already great :)
Afterwards, we can have a look how we can optimize it to give a good user experience. I try to get more familiar with annotation processing this weekend to find out how it could help.

**Comment by jobe-m** [comment]
> ... is the main problem with annotations the Java reflection part?

Yes, I guess so because `this.java.getAnnotation(T::class.java)` is not available and I do not get the annotations (AllOf, NoneOf...). Maybe I missed a Kotlin plugin or something similar that adds the "class" annotation objects to the system class.

Your suggestions sound good. Soywiz (the creator of Korge) also mentioned that he was playing with some Kotlin Compiler plugins which can generate code. I guess he meant exactly that and that these extentions can be controlled by annotations. I would be also eager to learn how that works. If you find a good source please let me know.

In the meantime I will get the current Fleks adaptation working with Korge and do a little Korge example using it :)

**Comment by Quillraven** [comment]
I played around a little bit with the annotation processing and [KotlinPoet](https://square.github.io/kotlinpoet/) and I was able to create e.g. a new class with a companion object that contains a unique property. I created two Component classes and with the annotation processing I created the same two components but each of them containing a companion object with a unique id field.

Quite interesting but the problem is that those classes must have a different name or different package. I was not able to modify the existing classes and from my understanding this is not the intention of annotation processing.

However, libraries like Lombok are doing that so it is somehow possible.

Unfortunately, I do not have a lot of time this month so I cannot experiment around more with that but it indeed looks promising :)

**Comment by jobe-m** [comment]
Sounds indeed promising. I also don't have too much time to look now into that and will concentrate to get my current state working. I got yesterday the basics of Fleks working with Korge. I oversaw the creation of the compmapper IDs which I got fixed after some debugging. Now the fun part starts. My anim objects (entities) get spawned and move over the screen. :)

Have you checked discord? I sent you a message there.

**Comment by Quillraven** [comment]
Sounds like you are making good progress, cool!

Sorry, I almost never check discord. I will have a look today but for the future, it is better to contact me here or by email: quillraven@gmail.com.

I mentioned you on a YouTube channel post that you help out with Fleks by the way ;)

**Comment by jobe-m** [comment]
I didn't know that you are a youtube master :)
Nice, I watched now both videos where you presented Fleks. And now I realized that by making Fleks compatible with Korge I removed the annotation API for defining Families which you praised most ;)

The current changed API feels a bit like a backstep and I don't know if you would like that. So what do you thing: should I create a PR to master or do you want it in a side branch?
Would you mind if we keep the changed version of Fleks in the Korge repo so that people can use it out ot the box with Korge?
At least until you have v2 with will have a nicer API and would work with Kotlin multiplatform and Korge.

**Comment by Quillraven** [comment]
I think for now we can release it as a separate version like 1.0-MPP or something.

In the future I want to introduce the annotation processing when I have time to get more familiar with it.

I will have a look at your changes in detail tomorrow and can give you better feedback then.

Thank you so far for everything !

**Comment by Quillraven** [comment]
@jobe-m: I updated now the github actions to also run the build workflow for the kmp branch. This workflow is currently failing because the unit tests do not run anymore. Can you please update them with the new API?

**Comment by jobe-m** [comment]
Great! I will check the unit test in a couple of days :)

**Comment by Quillraven** [comment]
@jobe-m the release should now be available on maven. Can you please try it in your KorGE test project if it works?

**Comment by Quillraven** [comment]
Thank you for trying it out and I think indeed we'd need to adjust the gradle setup: https://kotlinlang.org/docs/multiplatform-publish-lib.html

I am actually already using the `maven-publish` plugin but I think this is an important information:
```
When used with maven-publish, the Kotlin plugin automatically creates publications for each target 
that can be built on the current host, except for the Android target, which needs an additional 
step to configure publishing.
```

Most likely we need a block which defines the different targets. Right now this is missing as it was not necessary with the JVM target alone. Something like this might be the solution:

```
kotlin {
    jvm()
    js()
    mingwX64()
    linuxX64()
    val publicationsFromMainHost =
        listOf(jvm(), js()).map { it.name } + "kotlinMultiplatform"
    publishing {
        publications {
            matching { it.name in publicationsFromMainHost }.all {
                val targetPublication = this@all
                tasks.withType<AbstractPublishToMaven>()
                        .matching { it.publication == targetPublication }
                        .configureEach { onlyIf { findProperty("isMainHost") == "true" } }
            }
        }
    }
}
```

Unfortunately, I am quite busy at the moment and do not have time to play around with it but when I have it again then I will try. I  have almost no experience with KMP but I think we will then also have separate sourcesets for js/native/etc.? Are they then empty or do we need to do something? No idea :D

**Comment by jobe-m** [comment]
No problem. Take your time.

**Comment by Quillraven** [comment]
I will close this issue as it seems Fleks made it successfully into KorGE :) Awesome job @jobe-m !

https://github.com/korlibs/korge-next/pull/472#issuecomment-1070331178

**Comment by jobe-m** [comment]
+1
I will check korge-fleks with your master branch regularly to sync new features and additions. I also guess you can scrap the kmp branch (or mark it obsolete). In the readme on the master branch we can mention that the multiplatform version of Fleks is now available from Korge version 2.6.3 upwards.
